standalone-dns-proxy/cmd/standalonednsproxy.go

```diff
@@ -0,0 +1,578 @@
+	if err != nil {
+		log.WithError(err).Fatal("Failed to start DNS Proxy")
+		return err
+	}
+	log.Infof("DNS Proxy started on %s:%d", sdp.args.Address, sdp.args.Port)
+
+	// Create the cilium agent connection trigger
+	err = sdp.createciliumAgentConnectionTriggerTrigger()
+	if err != nil {
+		log.WithError(err).Error("Failed to create the trigger for connecting to Cilium agent")
+		return err
+	}
+
+	// trigger the cilium agent connection
+	sdp.ciliumAgentConnectionTrigger.TriggerWithReason("Start standalone DNS proxy")
+	return nil
+}
+
+// createciliumAgentConnectionTriggerTrigger creates a trigger to connect to the cilium agent
+// 1. It tries to connect to the cilium agent
+// 2. If the connection is successful, it tries to start the grpc streams
+// 3. If the streams are started, it tries to subscribe to the policy state as go routine
+func (sdp *StandaloneDNSProxy) createciliumAgentConnectionTriggerTrigger() error {
+	var err error
+	sdp.ciliumAgentConnectionTrigger, err = trigger.NewTrigger(trigger.Parameters{
+		Name:        "start-cilium-agent-connection",
+		MinInterval: 5 * time.Second,
+		TriggerFunc: func(reasons []string) {
+			defer func() {
+				if r := recover(); r != nil {
+					log.Errorf("Recovered from panic in trigger function: %v", r)
+				}
+			}()
+			log.Infof("Triggering cilium agent connection: %v", reasons)
+			// 1. Try creating the connection to the cilium agent
+			err := sdp.ConnectToCiliumAgent()
+			if err != nil {
+				log.WithError(err).Error("Failed to connect to cilium agent")
+				return
+			}
+
+			sdp.mu.Lock()
+			defer sdp.mu.Unlock()
+			// 2. Try starting the cilium agent connection
+			// only create the client if no stream is already open
+			// Imagine a scenarios where two triggers are fired at the same time
+			// and both try to create the client at the same time
+			// Due to the mutex, only one of them will create the client and start the stream
+			// The other one will just return
+			if sdp.policyStateStream == nil {
+				ctx, cancel := context.WithCancel(context.Background())
+
+				err = sdp.CreateClient(ctx)
+				if err != nil {
+					log.WithError(err).Error("Failed to create client")
+					cancel()
+					return
+				}
+
+				// 3. Try to subscribe to the policy state
+				sdp.cancelStreamPolicyStateStream = cancel // Store the cancel function for later use
+				go sdp.streamPolicyState(ctx)
+			}
+		},
+	})
+	if err != nil {
+		log.Errorf("Failed to create trigger: %v", err)
+		return err // Return the error after logging
+	}
+	return nil
+}
+
+// Note: isHost is always false as it is not used in the current implementation
+// TODO: Remove isHost from the function signature
+func (sdp *StandaloneDNSProxy) LookupEPByIP(ip netip.Addr) (ep *endpoint.Endpoint, isHost bool, err error) {
+	log.Debugf("LookupEPByIP: %s", ip.String())
+
+	// find the identity from the cache
+	secId, ok := sdp.ipToIdentityCache[ip.String()]
```

Comment:
what synchronisation guarantees that this map isn't accessed concurrently?

Reply:
Added the lock to make sure it is synchronized.  For some reason I thought I was using `sync.Map`.